Resolve vtereshkov#334 (Return user defined codes from Umka programs)

skejeton opened the pull request:
* And exit code parameter to `exit` (`exit(code: int)`)
* `main` can now return both `int` and `void`, returning a custom exit code and `0` exit code respectively.

vtereshkov reviewed (changes requested):
@skejeton Hic sunt dracones. You're entering a very dangerous land. Think twice before you proceed :)

1. It's a breaking change in Umka API -- let's invite @marekmaskarinec for review
2. If we accept it, we should fix:
  2.1. API docs
  2.2. Version number (to 1.4)
  2.3. 3D camera example
  2.4. tophat
3. What if `exit(code)` is called from an Umka function called by `umkaCall()` rather than `umkaRun()`? How do we separately return the error code and the function result? Should the function return anything any result in this case?
4. What if `exit(code)` is called from a fiber other than the main fiber? Should we terminate only this fiber or also the main fiber?
5. Should we suppress memory leak analysis on `exit(0)`? On `exit(1)`? In other words, do we consider this termination to be normal or abnormal?  (See `terminatedNormally`)
6. I discourage accessing `vm.mainFiber`'s registers directly from `umka_api.c`. In view of 4, it may be merely incorrect.
7. Code organization and style:
  7.1. As `exit()` now performs a nontrivial operation, I'd propose writing a `doBuiltinExit()` function called from `doCallBuiltin()`
  7.2. I'd rename `int *result` in Umka API functions into `int *errorCode` to avoid confusion with `UmkaStackSlot *result` (see also 3)
  7.3. An opening `{` is put on a new line in my C code (as opposed to Umka code)

Questions 3 and 4 are why I haven't done this myself yet.

skejeton commented:
I'm not sure how the scope of this conflicts with 3D camera example, since I kept `main` with `void` compatible, nor does it use `umkaRun`.

vtereshkov commented:
@skejeton 

4, 6. You put the error code to the current fiber's result register:
```
fiber->reg[VM_REG_RESULT] = *fiber->top;
```
and then read the main fiber's result register:
```
*result = comp->vm.mainFiber->reg[VM_REG_RESULT].intVal;
```
Why do you think this is the same fiber?

And of course, it's better to hide the register access in `compilerRun()`, as the `umka*` functions are the highest level abstractions that ideally shouldn't know anything about the VM fibers and registers.

> I'm not sure how the scope of this conflicts with 3D camera example, since I kept main with void compatible, nor does it use umkaRun.

But is does use `umkaCall()` that should perhaps also be modified to return an optional error code,

skejeton commented:
I fixed 7 and 1 except for tophat (I will do it if this is merged, since this isn't in the main branch right now) and 3D example (I didn't find anything).

vtereshkov commented:
@marekmaskarinec What do you think an `exit(code)` should do in a function called from `umkaCall()`?

skejeton commented:
I'm also now confused on why whether the program terminated normally or not should affect if the memory leaks are displayed or not.

vtereshkov commented:
> I'm also now confused on why whether the program terminated normally or not should affect if the memory leaks are displayed or not.

An abnormal termination means that the program has not correctly exited all the scopes associated with the `{...}` blocks, so we can be 99.9 % sure it will have a lot of memory leaks. To display them makes no sense.

skejeton commented:
Ah I see what you mean. I'm going to modify the behavior of `OP_HALT` to pull last value from the stack, since if I emit `BUILTIN_EXIT` before the program ends, the `OP_HALT` will never execute, and alternatively, if I do `BUILTIN_EXIT` after `OP_HALT`, the instruction listing will not include the `BUILTIN_EXIT` because it prints instructions up until `OP_HALT`.

vtereshkov commented:
@skejeton @marekmaskarinec Another question to discuss: 

Both `umkaRun()` and `umkaCall()` already return a `bool` value to show if the program has terminated with a runtime error (including an `error()` call). 

Should we treat an `exit(1)` as something fundamentally different from an `error("OMG")`? Or maybe we can just add an (optional) error code to `error()` and change the return types of  `umkaRun()` and `umkaCall()` from `bool` to `int`? In this case, we don't need to modify `exit()`.

skejeton commented:
I thought about this too, however the user might not want to show an error when exiting with a parameter.

skejeton commented:
I also think it's better to return `false` if we keep the boolean return value, on an exit call that returns a nonzero value.

vtereshkov commented:
> I thought about this too, however the user might not want to show an error when exiting with a parameter.

We can make the message string optional. If it's omitted, no message is displayed.

A bigger problem is again with compatibility. The new `umkaRun`/`umkaCall` behavior will perhaps imply that the "normal" exit will produce a zero result (implicitly converted to `false`). So we should carefully invert our logic everywhere.

skejeton commented:
I pushed the change, in this case the `exit` function now works fiber-wise, there's no check for `mainFiber` in `umkaRun`, instead I modified `compilerCall` and `compilerRun` to return the exit codes of the fiber. 

I changed the behaviour of `umkaRun` and `umkaCall` to return false on a faulty exit code, however this might not be a great idea, and it might better to change `error`, as you proposed. I want to hear your opinion on this change, and if you think changing `error` would be a good idea I'll work on it.

vtereshkov commented:
Let's listen to Marek. It was his request, so I'd like to know if he is OK with returning error codes from `error()` (which means `handlerRuntime` for us).

skejeton commented:
Yeah I think so too, I'm going to budge him on the server tomorrow if he doesn't respond by then.
In my view - I think the `error` change is the most logical, the only edge case I can think of is would `error(0)` be the same as `exit()`? Maybe in that case it's a wise idea to keep the boolean to distinguish between those two.

vtereshkov commented:
Hmm, I finally understand what I actually wanted to say. For me, the question is not what to choose, `error` or `exit`, but whether we can handle returning an error code just as a runtime error (with `handlerRuntime`). If we can, it will greatly simplify everything.

skejeton commented:
What is `handlerRuntime`? I did a search on the repository and couldn't find it.

vtereshkov commented:
Ah, I already renamed it to a more appropriate `runtimeHandler`.

skejeton commented:
The only problem I can think of your solution is that the exit code still needs to be returned from `main`, but we can't call the `runtimeHandler` to set the exit code, since that wouldn't end the program gracefully.

vtereshkov commented:
We needn't copy C here and return anything from `main`. We can say, if `main` returns normally, it's zero. If you need a non-zero, explicitly call `exit` or `error`.

skejeton commented:
I agree.

skejeton commented:
Can't seem to figure out how to push an empty string literal. I'll figure it out tomorrow.

skejeton commented:
I've implemented the `error` proposal, note that I'm still returning `bool` from `umkaRun` and `umkaCall`, because I'm not sure how to handle errors with code 0. If the error is passed code 0 then `umkaRun`/`umkaCall` will return 0, indicating no error, which is incorrect.

vtereshkov commented:
> If the error is passed code 0 then umkaRun/umkaCall will return 0, indicating no error, which is incorrect.

Why is it incorrect? I think `exit(0)` is a no-error termination.

(Except that if a function called with `umkaCall()` has exited on its own `return`, you can use `umkaCall()` again, but if it has exited on a `exit(0)`, you cannot, and these states are indistinguishable, as `umkaCall()` returns 0 in both cases. However, I don't think this is very important.)

skejeton commented:
I see. Then would the error message on `error(0)` be displayed, and also, if the message parameter is empty, would the error message be displayed, and how?

vtereshkov commented on src/umka_vm.c:
```diff
@@ -248,7 +248,7 @@ static FORCE_INLINE HeapPage *pageFind(HeapPages *pages, void *ptr, bool warnDan
 
             if (warnDangling && chunk->refCnt == 0)
-                pages->error->runtimeHandler(pages->error->context, "Dangling pointer at %p", ptr);
+                pages->error->runtimeHandler(pages->error->context, 1, "Dangling pointer at %p", ptr);
```
@skejeton Let's define something like a `VM_RUNTIME_ERROR = 1` in `umka_vm.h` and use it here.

skejeton commented:
Ah I see the error is printed only if the code isn't 0, still unsure on what to do in case of the message being empty.

vtereshkov commented on src/umka_expr.c:
```diff
@@ -1353,8 +1353,21 @@ static void parseBuiltinErrorCall(Compiler *comp, Type **type, Const *constant)
+
+    if (comp->lex.tok.kind == TOK_RPAR)
+    {
+        genPushGlobalPtr(&comp->gen, "");
```
@skejeton An Umka string is not a C string. Please use `storageAddStr(&comp->storage, 0)`.

vtereshkov commented on doc/lang.md:
````diff
@@ -804,13 +804,13 @@ Checks whether the `child` fiber function has not yet returned.
 
 ```
-fn error(msg: str)
+fn error(code: int, msg: str = "")
````
@skejeton If the error code is not optional, you'll have to fix all the occurences of `error()` (including Umka examples and tophat).

marekmaskarinec replied:
I don't think error should allow you to specify the return code. It should act just like any other runtime error.

vtereshkov commented:
@skejeton Now I don't see any difference between `exit()` and `error(0)`. Do we need both?

skejeton commented:
I thin `exit` can be removed, though `error(0)` is a bit confusing even if practically equivalent, the only difference is that `exit`, currently exits out of a fiber while `error(0)` exits to the closest setjmp.

skejeton commented:
I think merging into `exit` works. Asking again, but if the `msg` is empty, should we output the stack trace, line info and other info or not? I think if someone writes a utility program that's very unnecessary clogging of stderr/out.

marekmaskarinec commented:
> @marekmaskarinec What do you think an `exit(code)` should do in a function called from `umkaCall()`?

What about switching to an integer return value and returning values in the libc fashion? A negative value would mean a runtime error. A positive value (or zero) would be the return code.

vtereshkov commented:
> Asking again, but if the msg is empty, should we output the stack trace, line info and other info or not? I think if someone writes a utility program that's very unnecessary clogging of stderr/out.

@skejeton I think the end user can decide what to do. He/she will get an error code from `umkaRun`/`umkaCall`, request the `UmkaError` structure and analyze them (maybe in a fashion proposed by @marekmaskarinec).

vtereshkov commented on examples/3dcam/3dcam.c:
```diff
@@ -92,7 +92,7 @@ int main(void)
                 BeginMode3D(camera);
 
-                bool umkaOk = umkaCall(umka, umkaDrawBodies, 0, NULL, NULL);
+                bool umkaOk = umkaCall(umka, umkaDrawBodies, 0, NULL, NULL, NULL);
```
Remove the extra `NULL` and add ` == 0`

skejeton replied:
Fixed.

vtereshkov commented on runtime/std.um:
```diff
@@ -2,11 +2,13 @@
 
 // Memory
 
+const runtimeError* = 1
```
Should it be documented?

skejeton replied:
Yes. Fixed.

skejeton commented:
I think it's implemented now.

vtereshkov commented on src/umka_api.c:
```diff
@@ -61,6 +61,7 @@ static void runtimeError(void *context, const char *format, ...)
     strcpy(comp->error.fileName, debug->fileName);
     strcpy(comp->error.fnName, debug->fnName);
     comp->error.line = debug->line;
+    comp->error.code = code;
```
Maybe also add `comp->error.code = 1` to the compile-time error handler?

skejeton replied:
Fixed.

vtereshkov commented on src/umka_vm.h:
```diff
@@ -126,8 +130,7 @@ typedef enum
 
     // Misc
-    BUILTIN_EXIT,
-    BUILTIN_ERROR
```
Please also remove the `"error"` item from the function name spellings in `umka_vm.c`.

skejeton replied:
Fixed.

vtereshkov reviewed (changes requested):
I begin to like it.

vtereshkov commented:
@skejeton If Marek approves this, we'll also need to rebuild the playground. If you cannot do it, I'll do it myself after this pull request is merged.

skejeton commented:
I rebuilt the playground, I also noticed that there's an error in `umka.c` that we didn't catch, I fixed it.

marekmaskarinec commented:
> @marekmaskarinec Do you agree with the proposed changes?
> 
>     * `exit()` accepts an error code and an optional error message
> 
>     * `error()` removed
> 
>     * `umkaRun()` and `umkaCall()` return `int`, i.e., the error code (0 is no error)

How does one distinguish between a runtime error and a code passed to `exit`?

BTW @skejeton we will need to figure out how to implement this in tophat. Currently `exit` isn't used because it would just quit the current `umkaCall`, so we have `window.quit` instead.

vtereshkov commented:
> How does one distinguish between a runtime error and a code passed to exit?

@marekmaskarinec @skejeton We can redefine `VM_RUNTIME_ERROR` (and `std.runtimeError`?) to -1.

skejeton commented:
> How does one distinguish between a runtime error and a code passed to `exit`?

If no message is provided it's not printed in the default runtime (umka.c).

skejeton commented:
> BTW @skejeton we will need to figure out how to implement this in tophat. Currently exit isn't used because it would just quit the current umkaCall, so we have window.quit instead.

We just use the returned value from `umkaCall` in `frame`

marekmaskarinec commented:
> > How does one distinguish between a runtime error and a code passed to `exit`?
> 
> If no message is provided it's not printed in the default runtime (umka.c). So `exit(code)` is equivalent to C's `exit`.

But there are also runtime errors generated by umka itself.

skejeton commented:
> > > How does one distinguish between a runtime error and a code passed to `exit`?
> > 
> > 
> > If no message is provided it's not printed in the default runtime (umka.c). So `exit(code)` is equivalent to C's `exit`.
> 
> But there are also runtime errors generated by umka itself.

Gotcha, I made runtime errors negative.

vtereshkov commented:
>  But there are also runtime errors generated by umka itself.

@marekmaskarinec All runtime errors produced by the Umka VM and standard library will return error code -1. Your custom codes in `exit` can be any integers. Is it what you're expecting?
